### DISCUSSION TITLE
Documentation

### DISCUSSION THREAD

**Discussion by drusellers** [quartznet/quartznet#1796] [Ideas] [answered]
Before I go and spend time on this, I'd like to discuss some possible work I would be happy to do to help me out and possibly improve the documentation. 

# First, Restructure the tutorial section

I have enjoyed the model laid out by [Divio](https://documentation.divio.com/) and so i would first like to explore that model. To that end, I would break up the tutorial section into a few tutorials covering the basic usage. But would pull chunks out into an explanation area that addresses specific topics. Notably, try and break down `Jobs and Triggers`, `More about jobs & job details`, and `more about triggers`. Those seem less like tutorials and more about explaining the system.

# Second, add more how tos

I'd love to see some structured how tos on "writing a one off job", "write a job that fires every friday", etc.
Also, maybe a document or two on dealing with failed jobs (something I just went through)

---

Anyways, I feel like those would be good starting points and would definitely help out future-me when I'm digging back into Quartz.

**Comment by lahma** [comment, marked as answer]
This would definitely be much appreciated! Documentation is always lacking and as you've noticed, it hasn't evolved much from the early days. Most additions have been documenting the DI extensions etc. So such small walkthroughs/howtos sound like a great addition!

What I gather, you are also a native English speaker so that might improve the readability compared to documentation made by a person like me who might not have such active vocabulary to write documentation that keeps reader interested 😉

**Reply by drusellers** [reply]
Ok, excellent. I'll start putting some items together. My goal will to write a few How To's that I would like to have you validate first. :) and from there we can find a rhythm. 

Also, FWIW - you are doing amazingly well with your English. I had no clue.

**Comment by JasonLandbridge** [comment]
@drusellers, thanks for taking the time do improve the documentation! I kinda have a question, I just spend way to much time trying to figure out how to setup my database (sqlite) with the tables that are provided [here](https://github.com/quartznet/quartznet/blob/main/database/tables/tables_sqlite.sql) while doing it code first with entity framework. After a few hours I figured out the solution but it is quite a bit of code, but, it is a method that will work with every database that works with entity framework so it would be quite valuable.

My question, should I make an example project (which is kinda hard to figure out) or write a How-to guide referencing a folder with the code files and steps to set it up? Would you have a preference for a particular page?

**Reply by drusellers** [reply]
@JasonLandbridge ok, to make sure I'm understanding correctly. You wrote C# objects and their respective migrations `.cs` files? (If you had a repo with the output that would be super helpful to see). Now that you have figured this out, how do you share it with the rest of the quartz user base? 

If it is a collection of `.cs` files containing Entity Framework classes and then the appropriate migrations, and if it is a small number of files you could start by embedding them right into the docs and include a "How to use". However, my guess is that it is NOT just a few files.

Ok, so now this is more of a mapping your idea to my own brain type question. Is the goal of this to provide an easy way to query the data? By easy, I mostly mean "comfortable" in that I get to query (aka read only) with EF Core and LINQ. If so, I think a PR with a project that provides EF Core models on TOP of "SQL Views" could be really interesting. Because, I think writing should only be done through the Quartz APIs. In that case I'd personally enjoy a project added to the solution that includes the EF Core mappings + a migration that creates views on top of the existing tables.

Another contextual point that I personally have, is that trying to get the tables working with "code first" is not the correct approach. That you would instead add a custom migration to your project that adds the tables. That would be a solid piece of documentation / how to that probably does need to be written since I too got tripped up by it.

Ok, I'm sure I've only made things worse, but if you can help expand my understanding of your goals, I'd be happy to help direct you further.

**Reply by JasonLandbridge** [reply]
Thanks for the extensive reply! I realize my post was quite vague, I blame it on a long day of programming :sweat_smile:
> (If you had a repo with the output that would be super helpful to see). Now that you have figured this out, how do you share it with the rest of the quartz user base?

 I forgot to mention that I used the Entity configurations from [AppAny.Quartz.EntityFrameworkCore.Migrations](https://github.com/appany/AppAny.Quartz.EntityFrameworkCore.Migrations) so naturally a small note on just installing this and the usage of this might be an idea. Currently it doesn't support Sqlite out of the box, but I have raised [an issue](https://github.com/appany/AppAny.Quartz.EntityFrameworkCore.Migrations/issues/47) there where I would like to implement it. 

> Ok, so now this is more of a mapping your idea to my own brain type question. Is the goal of this to provide an easy way to query the data? By easy, I mostly mean "comfortable" in that I get to query (aka read only) with EF Core and LINQ. If so, I think a PR with a project that provides EF Core models on TOP of "SQL Views" could be really interesting. Because, I think writing should only be done through the Quartz APIs. 

The intention is to write a guide on how to setup the database (create the needed tables), which is then used as a jobstore for Quartz and configure it correctly. Maybe some configuration samples per database type? As I have understood it as well, you should never directly query/write to the tables created by Quartz. It should only be interacted with through the interfaces provided by IScheduler etc. 

> Another contextual point that I personally have, is that trying to get the tables working with "code first" is not the correct approach. That you would instead add a custom migration to your project that adds the tables. That would be a solid piece of documentation / how to that probably does need to be written since I too got tripped up by it.

I completely agree, that is what the principle behind it is. So as an example: 

```cs 
public class DbContext {

  protected override void OnModelCreating(ModelBuilder modelBuilder)
  {
    // Adds Quartz.NET Sqlite tables to EntityFrameworkCore
    modelBuilder.AddQuartz(builder => builder
      .UseSqlite()
      .UseSchema("quartz")
      .UseNoPrefix());
  }
}

```

The above would add all the tables required by Quartz through the use of the model builder like this:

```cs
public class QuartzSchedulerStateEntityTypeConfiguration : IEntityTypeConfiguration<QuartzSchedulerState>
{

    public void Configure(EntityTypeBuilder<QuartzSchedulerState> builder)
    {
        builder.ToTable($"{QuartzDatabaseConfig.Prefix}SCHEDULER_STATE");

        builder.HasKey(x => new { x.SchedulerName, x.InstanceName });

        builder.Property(x => x.SchedulerName)
            .HasColumnName("sched_name")
            .HasColumnType("text")
            .IsRequired();

        builder.Property(x => x.InstanceName)
            .HasColumnName("instance_name")
            .HasColumnType("text")
            .IsRequired();

        builder.Property(x => x.LastCheckInTime)
            .HasColumnName("last_checkin_time")
            .HasColumnType("bigint")
            .IsRequired();

        builder.Property(x => x.CheckInInterval)
            .HasColumnName("checkin_interval")
            .HasColumnType("bigint")
            .IsRequired();
    }
}

```

This would create the table with the needed properties, while also setting up the correct relationships. The cool thing here is that by doing it this way, the tables are never directly exposed through the DBContext. As would be the case if I have a property such as: 
```cs 
public DbSet<RandomTable> RandomTables { get; set; }
```

These steps are done by [AppAny.Quartz.EntityFrameworkCore.Migrations](https://github.com/appany/AppAny.Quartz.EntityFrameworkCore.Migrations) in the background so after a simple setup like this:
```cs
public class DbContext {

  protected override void OnModelCreating(ModelBuilder modelBuilder)
  {
    // Adds Quartz.NET Sqlite tables to EntityFrameworkCore
    modelBuilder.AddQuartz(builder => builder
      .UseSqlite()
      .UseSchema("quartz")
      .UseNoPrefix());
  }
}

```

You then run `dotnet ef add AddQuartzTables` to create the migration and then you're done. The tables are setup correctly and will be part of the migration process. This should in theory work for every database that Entity Framework supports on a code-first basis. 

## "But wait there is more!"

I also got tripped up by my usage of [Autofac.Extras.Quartz](https://github.com/alphacloud/Autofac.Extras.Quartz), since I both use Autofac and Quartz so they have to play nice together. The way I had to configure all of the above while also using `Autofac.Extras.Quartz` was the following:

```cs
using Autofac;
using Autofac.Extras.Quartz;

public class BackgroundServicesModule : Autofac.Module
{
    protected override void Load(ContainerBuilder builder)
    {
        var assembly = Assembly.GetExecutingAssembly();

        var quartzProps = new NameValueCollection
        {
            { "quartz.scheduler.instanceName", "PlexRipper Scheduler" },
            { "quartz.serializer.type", "json" },
            { "quartz.threadPool.type", "Quartz.Simpl.SimpleThreadPool, Quartz" },
            { "quartz.threadPool.threadCount", "10" },
            { "quartz.jobStore.type", "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz" },
            { "quartz.jobStore.misfireThreshold", "60000" },
            { "quartz.jobStore.lockHandler.type", "Quartz.Impl.AdoJobStore.UpdateLockRowSemaphore, Quartz" },
            { "quartz.jobStore.dataSource", "default" },
            { "quartz.jobStore.tablePrefix", QuartzDatabaseConfig.Prefix },
            { "quartz.jobStore.useProperties", "true" },
            { "quartz.jobStore.driverDelegateType", "Quartz.Impl.AdoJobStore.SQLiteDelegate, Quartz" },
            { "quartz.dataSource.default.provider", "SQLite-Microsoft" },
            { "quartz.dataSource.default.connectionString", PathProvider.DatabaseConnectionString }
        };

        // Source: https://github.com/alphacloud/Autofac.Extras.Quartz/blob/develop/src/Samples/Shared/Bootstrap.cs
        builder.Register(_ => new ScopedDependency("global"))
            .AsImplementedInterfaces()
            .SingleInstance();

        // Register Quartz dependencies
        builder.RegisterModule(new QuartzAutofacFactoryModule
        {
            JobScopeConfigurator = (cb, tag) =>
            {
                // override dependency for job scope
                cb.Register(_ => new ScopedDependency("job-local " + DateTime.UtcNow.ToLongTimeString()))
                    .AsImplementedInterfaces()
                    .InstancePerMatchingLifetimeScope(tag);
            },
            ConfigurationProvider = _ => quartzProps,
        });

        // SyncServerJob and DownloadProgressJob are Job types that inherrit IJob
        builder.RegisterModule(new QuartzAutofacJobsModule(typeof(SyncServerJob).Assembly));
        builder.RegisterModule(new QuartzAutofacJobsModule(typeof(DownloadProgressJob).Assembly));


        // register all I*Scheduler
        builder.RegisterAssemblyTypes(assembly)
            .Where(t => t.Name.EndsWith("Scheduler"))
            .AsImplementedInterfaces()
            .SingleInstance();
    }
}

```

I had to browse through many GitHub projects in the hope someone had my similar setup with a working config. Eventually I found it and then it started working correctly. So maybe a documentation section with different setups, and how to configure for them would be nice. The https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html page was however very nice in that it described every config option in detail. 

So, yeah I hope my ramblings make sense now :smile:

**Reply by drusellers** [reply]
Ok, I have a _much_ better understanding of what you are trying to achieve, and I really appreciate that you are willing to take the time to give back. I do think having a page that points out this package makes a lot of sense. I'm thinking that the first version of the document could live in `quartz-3.x/configuration` for the time being. Since this feels like a configuration item. That might even spur me to write a bit more getting started around migrations. Hmmm. 🤔 

On this page, a call out about SQLite is worth while (a link to the issue for others), but I'd suggest putting it towards the bottom as I suspect SQLite is not a common choice, but its good to have the note!

Additionally, a call out to the Autofac item that tripped you up would be a good note as well.

**Reply by JasonLandbridge** [reply]
Awesome and my pleasure, I'm incredibly grateful for Quartz and the already good documentation so I'm happy to give back!

So as a start, this is what I would like to propose:


1. Entity Framework integration
	-  In [docs/documentation/quartz-3.x/packages](https://github.com/quartznet/quartznet/tree/main/docs/documentation/quartz-3.x/packages) a new page called `Entity Framework Integration` that would reference the  [AppAny.Quartz.EntityFrameworkCore.Migrations](https://github.com/appany/AppAny.Quartz.EntityFrameworkCore.Migrations) package and how to set it up. I could also describe the manual way in both a code-first and database-first way. For the database way I will reference the [SQL scripts](https://github.com/quartznet/quartznet/blob/main/database/tables/). This page could then be referenced in the [Creating and initializing database](https://www.quartz-scheduler.net/documentation/quartz-3.x/quick-start.html#creating-and-initializing-database) section 
2. Autofac integration
	-  In [docs/documentation/quartz-3.x/packages](https://github.com/quartznet/quartznet/tree/main/docs/documentation/quartz-3.x/packages) a new page called `Autofac Integration` that would reference the [Autofac.Extras.Quartz](https://github.com/alphacloud/Autofac.Extras.Quartz) package and how to set it up. This would include some config examples with different databases and a link to [configuration/reference.html](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html) for more configuration options.
	-  I would also like to add a "warning" at the top of the [aspnet-core-integration.html](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/aspnet-core-integration.html#installation), [hosted-services-integration.html](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/hosted-services-integration.html#using) and [microsoft-di-integration.html](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/microsoft-di-integration.html#installation) pages that when using Autofac not to use these integrations as they do not do anything in combination with Autofac. At least, that is what I believe because without any of these steps, I managed to get it working. But initially, while following these guides, it gave me a false sense of everything working while none of my configurations were picked up because everything reverted to default config.

I thought about adding this to the [Quartz.NET Configuration Reference](https://www.quartz-scheduler.net/documentation/quartz-3.x/configuration/reference.html) as you mentioned, but I'm afraid that it's hard to find the above, quite crucial, setup information. And by placing these both here: 
![image](https://user-images.githubusercontent.com/15127381/206864377-f5af588e-2814-4d83-891f-35f22531f056.png)

It would be more easily spotted, anyway let me know what you think :+1:
